plugins/out_azure_logs_ingestion/azure_logs_ingestion.c

```diff
@@ -0,0 +1,445 @@
+            msgpack_pack_str_body(&mp_pck, time_formatted, s);
+        }
+        else {
+            /* Append the time value as millis.nanos */
+            t = flb_time_to_double(&tm);
+            msgpack_pack_double(&mp_pck, t);
+        }
+
+        /* Append original map k/v */
+        for (i = 0; i < map_size; i++) {
+            k = map.via.map.ptr[i].key;
+            v = map.via.map.ptr[i].val;
+
+            msgpack_pack_object(&tmp_pck, k);
+            msgpack_pack_object(&tmp_pck, v);
+        }
+        msgpack_sbuffer_write(&mp_sbuf, tmp_sbuf.data, tmp_sbuf.size);
+        msgpack_sbuffer_destroy(&tmp_sbuf);
+    }
+
+    record = flb_msgpack_raw_to_json_sds(mp_sbuf.data, mp_sbuf.size);
+    if (!record) {
+        flb_errno();
+        msgpack_sbuffer_destroy(&mp_sbuf);
+        msgpack_unpacked_destroy(&result);
+        return -1;
+    }
+
+    msgpack_sbuffer_destroy(&mp_sbuf);
+    msgpack_unpacked_destroy(&result);
+
+    *out_buf = record;
+    *out_size = flb_sds_len(record);
+
+    return 0;
+}
+
+/* Gets OAuth token; (allocates sds string) */
+flb_sds_t get_az_li_token(struct flb_az_li *ctx)
+{
+    int ret = 0;
+    char* token;
+    size_t token_len;
+
+    if (pthread_mutex_lock(&ctx->token_mutex)) {
+        flb_plg_error(ctx->ins, "error locking mutex");
+        return NULL;
+    }
+    /* Retrieve access token only if expired */
+    if (flb_oauth2_token_expired(ctx->u_auth) == FLB_TRUE) {
+        flb_plg_debug(ctx->ins, "token expired. getting new token");
+        /* Clear any previous oauth2 payload content */
+        flb_oauth2_payload_clear(ctx->u_auth);
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "grant_type", 10, 
+                                        "client_credentials", 18);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "scope", 5, FLB_AZ_LI_AUTH_SCOPE,
+                                        sizeof(FLB_AZ_LI_AUTH_SCOPE) - 1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "client_id", 9,
+                                        ctx->client_id, -1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "client_secret", 13,
+                                        ctx->client_secret, -1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        token = flb_oauth2_token_get(ctx->u_auth);
+
+        /* Copy string to prevent race conditions */
+        if (!token) {
+            flb_plg_error(ctx->ins, "error retrieving oauth2 access token");
+            return NULL;
```

Comment:
This exit path doesn't seem to release the mutex which could cause a deadlock.

Reply:
Fixed